### PR TITLE
implementing new features in value expression editor

### PR DESCRIPTION
**Pull request by sudarshan0212**

### CODE REVIEW

**Comment by ambarkhan** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 2
           

Complexity increasing per file
==============================
- Ginger/Ginger/UserControlsLib/TextEditor/ValueExpression/ValueExpressionFlowDetailsEditorPage.xaml.cs  20
- Ginger/Ginger/UserControlsLib/ImageMakerLib/ImageMakerControl.xaml.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765&bid=20636913)

File: Ginger/Ginger/UserControlsLib/TextEditor/ValueExpression/ValueExpressionFlowDetailsEditorPage.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            if (WorkSpace.Instance.RunsetExecutor != null)
+            {
+                runset = WorkSpace.Instance.RunsetExecutor.RunSetConfig;
+            }
+            if (runset != null)
+            {
+                runner = WorkSpace.Instance.RunsetExecutor.Runners.Where(x => x.BusinessFlows.Where(bf => mContext.BusinessFlow != null && bf.Name == mContext.BusinessFlow.Name).FirstOrDefault() != null).FirstOrDefault();
+            }
+            RepositoryItemBase objtoEval = null;
+            switch (mObj)
```
Comment (ambarkhan) [inline Ginger/Ginger/UserControlsLib/TextEditor/ValueExpression/ValueExpressionFlowDetailsEditorPage.xaml.cs:64]: Codacy found an issue: [Add a 'default' clause to this 'switch' statement.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs [hunk 2]
```diff
@@ -396,10 +475,14 @@ private void AddSecurityConfiguration()
...
+                    ValueExpressionFunctionCategory Category = Attribute.GetCustomAttribute(mi, typeof(ValueExpressionFunctionCategory), false) as ValueExpressionFunctionCategory;
+                    ValueExpressionFunctionSubCategory SubCategory = Attribute.GetCustomAttribute(mi, typeof(ValueExpressionFunctionSubCategory), false) as ValueExpressionFunctionSubCategory;                    
                     if (token == null)
                         continue;
-
-                    AddWSSecurityConfig(tviSecuritySettings, desc.DefaultValue, expr.DefaultValue); //GetUTCDateTimeStamp());
+                    TreeViewItem Parent = AddOrGetCategory(Category.DefaultValue);
+                    TreeViewItem child = AddOrGetSubCategory(SubCategory.DefaultValue, Parent);
+                    child.Selected += HideHelp;
+                    AddWSSecurityConfig(child, desc.DefaultValue, expr.DefaultValue, SubCategory.DefaultValue); //GetUTCDateTimeStamp());
```
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:485]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

**Comment by ambarkhan** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- Ginger/Ginger/UserControlsLib/ImageMakerLib/ImageMakerControl.xaml.cs  2
- Ginger/Ginger/UserControlsLib/TextEditor/ValueExpression/ValueExpressionFlowDetailsEditorPage.xaml.cs  20
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765&bid=20636913)

**Comment by ambarkhan** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs  1
- Ginger/Ginger/UserControlsLib/ImageMakerLib/ImageMakerControl.xaml.cs  3
- Ginger/Ginger/UserControlsLib/TextEditor/ValueExpression/ValueExpressionFlowDetailsEditorPage.xaml.cs  20
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765&bid=20636913)

**Comment by ambarkhan** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 4
           

Complexity increasing per file
==============================
- Ginger/Ginger/UserControlsLib/TextEditor/ValueExpression/ValueExpressionFlowDetailsEditorPage.xaml.cs  17
- Ginger/Ginger/UserControlsLib/ImageMakerLib/ImageMakerControl.xaml.cs  3
- Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765&bid=20636913)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs [hunk 3]
```diff
@@ -463,40 +546,42 @@ private void AddVBSIfEval(TreeViewItem tviVars, string Desc, string Eval)
 
         private void AddEnvParams()
         {
+            TreeViewItem Parent = AddOrGetCategory("Data");
+            //TreeViewItem child = AddOrGetSubCategory("Environments", Parent);
```
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:550]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs [hunk 4]
```diff
@@ -168,9 +167,11 @@ private void BuildItemsTree()
 
         private void AddGlobalParameters()
         {
+            TreeViewItem Parent = AddOrGetCategory("Data");
+            //TreeViewItem child = AddOrGetSubCategory("Models Global Parameters", Parent);
```
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:171]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs [hunk 5]
```diff
@@ -575,16 +662,18 @@ private void InsertNewVarTreeItem(TreeViewItem parentTvi, VariableBase vb)
...
-            SetItemView(newVarTvi, "Add New String " + GingerDicser.GetTermResValue(eTermResKey.Variable) , varLevel, "@Add_16x16.png");
+            SetItemView(newVarTvi, "Add New String " + GingerDicser.GetTermResValue(eTermResKey.Variable) , varLevel, eImageType.Add);
             parentTvi.Items.Add(newVarTvi);
             newVarTvi.MouseDoubleClick += tviAddNewVarTreeItem_MouseDoubleClick;
         }
 
         private void AddDataSources()
         {
+            TreeViewItem Parent = AddOrGetCategory("Data");
+            //TreeViewItem child = AddOrGetSubCategory("Data Source", Parent);
```
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:673]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs [hunk 6]
```diff
@@ -537,20 +624,20 @@ private void AddVariables()
...
-                SetItemView(tviVars, GingerDicser.GetTermResValue(eTermResKey.BusinessFlow) + " " + GingerDicser.GetTermResValue(eTermResKey.Variables), "", "@Variable_16x16.png");
-                xObjectsTreeView.Items.Add(tviVars);
+                SetItemView(tviVars, GingerDicser.GetTermResValue(eTermResKey.BusinessFlow) + " " + GingerDicser.GetTermResValue(eTermResKey.Variables), "", eImageType.Variable);
+                child.Items.Add(tviVars);
 
                 foreach (VariableBase v in mContext.BusinessFlow.Variables.OrderBy("Name"))
                     InsertNewVarTreeItem(tviVars, v);
                 InsertAddNewVarTreeItem(tviVars, eVariablesLevel.BusinessFlow);
-                tviVars.IsExpanded = true;
+                //tviVars.IsExpanded = true;
```
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:633]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

**Comment by ambarkhan** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 4
           

Complexity increasing per file
==============================
- Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs  1
- Ginger/Ginger/UserControlsLib/ImageMakerLib/ImageMakerControl.xaml.cs  3
- Ginger/Ginger/UserControlsLib/TextEditor/ValueExpression/ValueExpressionFlowDetailsEditorPage.xaml.cs  19
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765&bid=20636913)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs (hunk 3 above)
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:550]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs (hunk 4 above)
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:171]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs (hunk 5 above)
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:673]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs (hunk 6 above)
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:633]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

**Comment by ambarkhan** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 4
           

Complexity increasing per file
==============================
- Ginger/Ginger/UserControlsLib/ImageMakerLib/ImageMakerControl.xaml.cs  3
- Ginger/Ginger/UserControlsLib/TextEditor/ValueExpression/ValueExpressionFlowDetailsEditorPage.xaml.cs  19
- Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765&bid=20636913)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs (hunk 3 above)
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:550]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs (hunk 4 above)
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:171]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs (hunk 5 above)
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:673]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs (hunk 6 above)
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:633]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

**Comment by ambarkhan** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 4
           

Complexity increasing per file
==============================
- Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs  7
- Ginger/Ginger/UserControlsLib/ImageMakerLib/ImageMakerControl.xaml.cs  3
- Ginger/Ginger/UserControlsLib/TextEditor/ValueExpression/ValueExpressionFlowDetailsEditorPage.xaml.cs  19
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765&bid=20636913)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs [hunk 7]
```diff
@@ -463,40 +565,42 @@ private void AddVBSIfEval(TreeViewItem tviVars, string Desc, string Eval)
 
         private void AddEnvParams()
         {
+            TreeViewItem Parent = AddOrGetCategory("Data");
+            //TreeViewItem child = AddOrGetSubCategory("Environments", Parent);
```
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:569]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs [hunk 8]
```diff
@@ -575,16 +681,18 @@ private void InsertNewVarTreeItem(TreeViewItem parentTvi, VariableBase vb)
...
-            SetItemView(newVarTvi, "Add New String " + GingerDicser.GetTermResValue(eTermResKey.Variable) , varLevel, "@Add_16x16.png");
+            SetItemView(newVarTvi, "Add New String " + GingerDicser.GetTermResValue(eTermResKey.Variable) , varLevel, eImageType.Add);
             parentTvi.Items.Add(newVarTvi);
             newVarTvi.MouseDoubleClick += tviAddNewVarTreeItem_MouseDoubleClick;
         }
 
         private void AddDataSources()
         {
+            TreeViewItem Parent = AddOrGetCategory("Data");
+            //TreeViewItem child = AddOrGetSubCategory("Data Source", Parent);
```
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:692]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs [hunk 9]
```diff
@@ -161,16 +160,19 @@ private void BuildItemsTree()
...
+            TreeViewItem dataRoot = AddOrGetCategory("Data");
+            dataRoot.IsExpanded = true;
         }
 
     
 
         private void AddGlobalParameters()
         {
+            TreeViewItem Parent = AddOrGetCategory("Data");
+            //TreeViewItem child = AddOrGetSubCategory("Models Global Parameters", Parent);
```
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:172]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs [hunk 10]
```diff
@@ -537,20 +643,20 @@ private void AddVariables()
...
-                SetItemView(tviVars, GingerDicser.GetTermResValue(eTermResKey.BusinessFlow) + " " + GingerDicser.GetTermResValue(eTermResKey.Variables), "", "@Variable_16x16.png");
-                xObjectsTreeView.Items.Add(tviVars);
+                SetItemView(tviVars, GingerDicser.GetTermResValue(eTermResKey.BusinessFlow) + " " + GingerDicser.GetTermResValue(eTermResKey.Variables), "", eImageType.Variable);
+                child.Items.Add(tviVars);
 
                 foreach (VariableBase v in mContext.BusinessFlow.Variables.OrderBy("Name"))
                     InsertNewVarTreeItem(tviVars, v);
                 InsertAddNewVarTreeItem(tviVars, eVariablesLevel.BusinessFlow);
-                tviVars.IsExpanded = true;
+                //tviVars.IsExpanded = true;
```
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:652]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

**Review by MeniKadosh1 — APPROVED** [review]

**Comment by ambarkhan** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 4
           

Complexity increasing per file
==============================
- Ginger/Ginger/UserControlsLib/ImageMakerLib/ImageMakerControl.xaml.cs  3
- Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs  7
- Ginger/Ginger/UserControlsLib/TextEditor/ValueExpression/ValueExpressionFlowDetailsEditorPage.xaml.cs  19
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765&bid=20636913)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs (hunk 7 above)
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:569]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs (hunk 8 above)
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:692]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs (hunk 9 above)
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:172]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs (hunk 10 above)
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:652]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

**Comment by ambarkhan** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 4
           

Complexity increasing per file
==============================
- Ginger/Ginger/UserControlsLib/TextEditor/ValueExpression/ValueExpressionFlowDetailsEditorPage.xaml.cs  19
- Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs  7
- Ginger/Ginger/UserControlsLib/ImageMakerLib/ImageMakerControl.xaml.cs  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765&bid=20636913)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs (hunk 7 above)
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:569]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs (hunk 8 above)
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:692]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs (hunk 9 above)
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:172]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

File: Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs (hunk 10 above)
Comment (ambarkhan) [inline Ginger/Ginger/ValueExpression/ValueExpressionEditorPage.xaml.cs:652]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6356765)

**Pull request merged**